src/main/java/nl/nn/testtool/TestTool.java

```diff
@@ -175,6 +182,25 @@ private Object checkpoint(String correlationId, String threadId,
 		}
 		return message;
 	}
+

```

Comment:
A loop within a loop doesn't seem necessary. In case you have 100 checkpoint you are looping 100 over 100 times (hence 10.000 times). Please also add comments why you are doing this. I would suggest to implement the method as follows. Please test it (as I didn't) and verify it is doing what you are trying to do.

```suggestion
	private void capCheckpointMessages(Report report) {
		List<Checkpoint> checkpointsToSkip = new ArrayList<Checkpoint>();
		Map<String, String> cappedMessages = new HashMap<String, String>();
		for (Checkpoint checkpoint : report.getCheckpoints()) {
			if (checkpoint.getMessage() != null && checkpoint.getMessage().length() > maxMessageLength) {
				if (cappedMessages.containsKey(checkpoint.getMessage())) {
					// For a message that is referenced by multiple checkpoints have one capped message that is
					// referenced by those checkpoints to prevent creating multiple string objects representing the
					// same string an occupying unnecessary memory.
					checkpoint.setMessage(cappedMessages.get(checkpoint.getMessage()));
				} else {
					String cappedMessage = checkpoint.getMessage().substring(0, maxMessageLength)
							+ "... ("+(checkpoint.getMessage().length() - maxMessageLength)+" more characters)";
					cappedMessages.put(checkpoint.getMessage(), cappedMessage);
					checkpoint.setMessage(cappedMessage);
				}
			}
		}
	}
```

Reply:
Using a <fullMessage,cappedMessage> map seems to take significantly more time. Note that any checkpoints that have their messages capped are removed from the list of checkpoints to iterate through.

For example, let's say we have a report with 100 checkpoints and 4 unique messages that should be capped. The outer loop would run 4 times. If iteration 1 were to replace the string for 40 checkpoints, iteration 2 would only loop through the 60 remaining checkpoints.

I've made some changes to my proposed solution to decrease the amount of loop iterations.